controllers/cache/deploy.go

```diff
@@ -0,0 +1,76 @@
+package cache
+
+import (
+	"context"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"log"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	harborCluster "src/github.com/goharbor/harbor-cluster-operator/api/v1"
+)
+
+func (d *defaultCache) Deploy(status *harborCluster.CRStatus) (*harborCluster.CRStatus, error) {
+	var actualCR *unstructured.Unstructured
+	var expectCR *unstructured.Unstructured
+	name := d.Request.Name
+	nameSpace := d.Request.Namespace
+	crdClient := d.DClient.Resource(virtualServiceGVR)
+
+	expectCR, err := d.generateRedisCR()
+	if err != nil {
+		return status, err
+	}
+	if err := controllerutil.SetControllerReference(d.Harbor, expectCR, d.Scheme); err != nil {
+		return status, err
+	}
+
+	actualCR, err = crdClient.Namespace(nameSpace).Get(name, metav1.GetOptions{})
+	if err != nil && errors.IsNotFound(err) {
+
+		if err := d.DeploySecret(); err != nil {
+			return status, err
+		}
+
+		d.Log.Info("Creating Redis.", "namespace", nameSpace, "name", name)
+		_, err = crdClient.Namespace(nameSpace).Create(expectCR, metav1.CreateOptions{})
+		if err != nil {
+			return status, err
+		}
+		status.Phase = harborCluster.DeployingPhase
+
+		d.Log.Info("Redis create complete.", "namespace", nameSpace, "name", name)
+	} else if err != nil {
+		return status, err
+	} else {
+		d.ExpectCR = expectCR
+		d.ActualCR = actualCR
+	}
+
+	return status, nil
+}
+
+// DeploySecret deploy the Redis Password Secret
+func (d *defaultCache) DeploySecret() error {
+	secret := &corev1.Secret{}
+	name := d.Request.Name
+	namespace := d.Request.Namespace
+	sc := d.generateRedisSecret(d.Labels)
+	if err := controllerutil.SetControllerReference(d.Harbor, sc, d.Scheme); err != nil {
+		return err
+	}
+	err := d.Client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: namespace}, secret)
```

Comment:
Use `context.TODO()` or sharing the context of harbor-cluster reconcile controller?

Reply:
ok, It has been modified, use the context of harbor-cluster reconcile controller.